### ISSUE TITLE
JavaScript syntax error with multi-line concise arrow functions

### ISSUE THREAD

**Issue by rluba** [sublimehq/Packages#222]
The following (valid) JavaScript code is considered invalid by the current version (d056ca6) of the JavaScript package:
```JavaScript
'use strict';

return functionA()
	.then(() => {
		functionB({
			propertyName: () => functionC()
				.then(() => {})
		});
	});
```

The last parenthesis is marked as invalid:
<img width="399" alt="js-error" src="https://cloud.githubusercontent.com/assets/4600573/13425996/f8457256-dfaa-11e5-85b9-cbb1b28bcb61.png">

I know the code itself doesn’t make much sense, but it’s valid JavaScript. The pattern occurred as part of a larger code segment. I’ve tried to reduce it as much as possible while still retaining the error.

**Comment by wbond** [comment]
I don't believe this is due to nesting arrow functions, but instead has to do with multi-line, "concise" arrow function bodies.

If you pull `.then()` up to the same line as `functionC()`, then it will highlight properly.

I will try to do some work on getting this handled properly. We are getting into a more and more stateful syntax for JS in order to handle different edge cases, and hopefully all of that state tracking can help us deal with something like this.

**Comment by wbond** [comment]
This should be resolved by e7a947c96bb65c38967ee25e5ed66ec8e8bdbcee

**Comment by rluba** [comment]
Yes, thanks. I can confirm that it’s working with the latest version.
